pinot-controller/src/main/java/com/linkedin/pinot/controller/api/upload/ZKOperator.java

```diff
@@ -0,0 +1,195 @@
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.linkedin.pinot.controller.api.upload;
+
+import com.linkedin.pinot.common.config.TableNameBuilder;
+import com.linkedin.pinot.common.metadata.segment.OfflineSegmentZKMetadata;
+import com.linkedin.pinot.common.metadata.segment.SegmentZKMetadataCustomMapModifier;
+import com.linkedin.pinot.common.metrics.ControllerMeter;
+import com.linkedin.pinot.common.metrics.ControllerMetrics;
+import com.linkedin.pinot.common.segment.SegmentMetadata;
+import com.linkedin.pinot.common.utils.FileUploadDownloadClient;
+import com.linkedin.pinot.controller.ControllerConf;
+import com.linkedin.pinot.controller.api.resources.ControllerApplicationException;
+import com.linkedin.pinot.controller.api.resources.FileUploadPathProvider;
+import com.linkedin.pinot.controller.helix.core.PinotHelixResourceManager;
+import com.linkedin.pinot.filesystem.PinotFS;
+import com.linkedin.pinot.filesystem.PinotFSFactory;
+import java.io.File;
+import java.net.URI;
+import javax.ws.rs.core.HttpHeaders;
+import javax.ws.rs.core.Response;
+import org.apache.helix.ZNRecord;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * The ZKOperator is a util class that is used during segment upload to set relevant metadata fields in zk. It will currently
+ * also perform the data move. In the future when we introduce versioning, we will decouple these two steps.
+ */
+public class ZKOperator {
+  private static final Logger LOGGER = LoggerFactory.getLogger(ZKOperator.class);
+  private PinotHelixResourceManager _pinotHelixResourceManager;
+  private ControllerConf _controllerConf;
+  private ControllerMetrics _controllerMetrics;
+
+  public ZKOperator(PinotHelixResourceManager pinotHelixResourceManager, ControllerConf controllerConf, ControllerMetrics controllerMetrics) {
+    _pinotHelixResourceManager = pinotHelixResourceManager;
+    _controllerConf = controllerConf;
+    _controllerMetrics = controllerMetrics;
+  }
+
+  public void completeSegmentOperations(SegmentMetadata segmentMetadata, URI finalSegmentLocationURI, File currentSegmentLocation, boolean enableParallelPushProtection, HttpHeaders headers,
+      FileUploadPathProvider provider) throws Exception {
+    String rawTableName = segmentMetadata.getTableName();
+    String offlineTableName = TableNameBuilder.OFFLINE.tableNameWithType(rawTableName);
+    String segmentName = segmentMetadata.getName();
+
+    ZNRecord znRecord = _pinotHelixResourceManager.getSegmentMetadataZnRecord(offlineTableName, segmentName);
+
+    // Brand new segment, not refresh, directly add the segment
+    if (znRecord == null) {
```

Comment:
May make it more readable if you break out the 'if' part into a method, and return ?